consumer/consumer_internal_test.go

```diff
@@ -251,6 +251,92 @@ func TestConsumerHandleMessagesMetricsReporting(t *testing.T) {
 	require.Equal(t, 1, mmh.ReportCount)
 }
 
+func TestConsumerHandleMessagesRetryOnError(t *testing.T) {
+	cfg := NewConfig("some-id")
+	c := Consumer{
+		config: &cfg,
+	}
+	metricsCh := make(chan *Metrics)
+	reportMetric := func(m Message, metrics *Metrics) {
+		metricsCh <- metrics
+	}
+	c.Metrics = metricsReporterFunc(reportMetric)
+
+	type msgHandleReq struct {
+		m     *Message
+		reply chan error
+	}
+	msgHandleCh := make(chan msgHandleReq)
+	handle := func(m *Message) error {
+		reply := make(chan error)
+		msgHandleCh <- msgHandleReq{m, reply}
+		return <-reply
+	}
+	c.Handle("topic", MessageConverterV1(NewConfig("")), HandlerFunc(handle))
+
+	msgCh := make(chan *sarama.ConsumerMessage, 1)
+	handleMessagesDone := make(chan struct{})
+	go func() {
+		defer close(handleMessagesDone)
+		c.handleMessages(
+			msgCh,
+			&mockOffsetStash{},
+			&mockHighWaterMarker{HighWaterMarkOffsetCount: 1},
+			"topic",
+			1,
+		)
+	}()
+
+	// Send a message to the channel; we should exponentially
+	// backoff until the message is handled OK. As the backoff is jittered
+	// (and controlled by the specific retry strategy chosen) we don't
+	// check the specific delays.
+	msgCh <- &sarama.ConsumerMessage{
+		Topic:  "topic",
+		Key:    []byte("key"),
+		Value:  []byte(`"body0"`),
+		Offset: 1,
+	}
+	for i := 0; i < 5; i++ {
+		select {
+		case req := <-msgHandleCh:
+			require.Equal(t, []byte("key"), req.m.Key.Bytes())
+			require.Equal(t, []byte(`"body0"`), req.m.Body.Bytes())
+			req.reply <- fmt.Errorf("some error")
+		case <-time.After(2 * time.Second):
+			t.Fatal("timed out waiting for handle request")
+		}
+	}
+	// Allow the handler to succeed.
+	select {
+	case req := <-msgHandleCh:
+		require.Equal(t, []byte("key"), req.m.Key.Bytes())
+		require.Equal(t, []byte(`"body0"`), req.m.Body.Bytes())
+		req.reply <- nil
+	case <-time.After(2 * time.Second):
+		t.Fatal("timed out waiting for handle request")
+	}
+
+	// We should receive a metrics report that
+	// the message was sent correctly.
+	select {
+	case m := <-metricsCh:
+		require.Equal(t, m.Attempts, 6)
+	case <-time.After(2 * time.Second):
+		t.Fatal("timed out waiting for handle request")
+	}
+
+	// When the channel is closed, the handler should terminate.
+	close(msgCh)
+	select {
+	case <-handleMessagesDone:
+	case <-time.After(2 * time.Second):
+		go panic("timed out waiting for handler loop to finish")
+		select {}
```

Comment:
What is the purpose of this line?

Reply:
nothing! it's a debugging remnant that needs to be removed :)
(FYI it's a useful idiom to use if your test hangs up and you want to know what all the other goroutines are doing, not just where the test code is).

Reply:
Oh cool, TIL again 🙏🏼

> and you want to know what all the other goroutines are doing

By using a third piece or just printing debug?

Reply:
By running panic in a separate goroutine, it means that the testing package can't catch it, so the exception will be caught by the standard runtime (which prints stack traces for all goroutines if `GOTRACEBACK=all`) instead of the testing package, which only prints one.

Reply:
Alright, thank you!